src/De/DeJson.jl

```diff
@@ -2,15 +2,359 @@ module DeJson
+            array_elements = T(undef, yyjson_arr_size(value_ptr))
+            for i in eachindex(array_elements)
+                elem_ptr = yyjson_arr_iter_next(iter_ptr)
+                array_elements[i] = deser(eltype(T), elem_ptr)
+            end
+        end
+        array_elements
+    end
+end
+
+function deser(::ArrayType, ::Type{T}, value_ptr::Ptr{YYJSONVal}) where {T<:Tuple}
+    return if yyjson_is_arr(value_ptr)
+        if T == Tuple
+            T(deser(Vector{Any}, value_ptr))
+        else
+            iter = YYJSONArrIter()
+            iter_ref = Ref(iter)
+            iter_ptr = Base.unsafe_convert(Ptr{YYJSONArrIter}, iter_ref)
+            GC.@preserve iter begin
+                yyjson_arr_iter_init(value_ptr, iter_ptr) || throw(YYJSONError("Failed to initialize array iterator."))
+                tuple_elements = Vector{Any}(undef, fieldcount(T))
+                for (i, type) in zip(eachindex(tuple_elements), fieldtypes(T))
+                    value_ptr = yyjson_arr_iter_next(iter_ptr)
+                    tuple_elements[i] = deser(type, value_ptr)
+                end
+            end
+            T(tuple_elements)
+        end
+    end
+end
+
+# CustomType
+
+function deser_arr(::CustomType, ::Type{T}, value_ptr::Ptr{YYJSONVal}) where {T}
+    iter = YYJSONArrIter()
+    iter_ref = Ref(iter)
+    iter_ptr = Base.unsafe_convert(Ptr{YYJSONArrIter}, iter_ref)
+    GC.@preserve iter begin
+        yyjson_arr_iter_init(value_ptr, iter_ptr) || throw(YYJSONError("Failed to initialize array iterator."))
+        type_elements = Vector{Any}(undef, fieldcount(T))
+        for (i, field_type) in zip(eachindex(type_elements), fieldtypes(T))
+            field_ptr = yyjson_arr_iter_next(iter_ptr)
+            value = field_ptr == C_NULL ? nulltype(field_type) : deser(T, field_type, field_ptr)
+            type_elements[i] = isempty(T, value) ? nulltype(field_type) : value
+        end
+        return T(type_elements...)
+    end
+end
+
+function typeof_yyjson_val(val_ptr::Ptr{YYJSONVal})
+    return if yyjson_is_str(val_ptr)
+        String
+    elseif yyjson_is_raw(val_ptr)
+        String
+    elseif yyjson_is_real(val_ptr)
+        Float64
+    elseif yyjson_is_int(val_ptr)
+        Int
+    elseif yyjson_is_bool(val_ptr)
+        Bool
+    elseif yyjson_is_obj(val_ptr)
+        Dict{String,Any}
+    elseif yyjson_is_arr(val_ptr)
+        Vector{Any}
+    elseif yyjson_is_null(val_ptr)
+        Nothing
+    else
+        Any
+    end
+end
+
+function check_types(::Type{E}, ::Type{T}) where {E,T}
+    return if E isa Union
+        any(x -> x <: T, Base.uniontypes(E))
+    else
+        <:(E, T)
+    end
+end
+
+function eldeser(::Type{T}, ::Type{E}, key::Union{AbstractString,Symbol}, value_ptr::Ptr{YYJSONVal}) where {T,E}
+    return try
+        if yyjson_is_str(value_ptr) && check_types(E, AbstractString)
+            unsafe_string(yyjson_get_str(value_ptr))
```

Comment:
@gryumov 
This function check that expected type (field) corresponds to `value_ptr` type